### PR TITLE
cockroachdb: introduce multi-register test

### PR DESCRIPTION
**Pull request by nvb**
This change introduces a new "multi-register" jepsen test for CockroachDB. The test is similar to "register" in that it performs reads and writes on atomic registers. However, unlike "register", the new test runs transactions that access multiple registers. It then uses the new Knossos "multi-register" checker to search for non-linearizable histories across the component transactions.

The goal here is to tickle single-key linearizability violations that only occur with multi-Range transactions. To that end, the test splits each register into its own table.

### CODE REVIEW

File: cockroachdb/src/jepsen/cockroach/multiregister.clj [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+(defn id->table
+  "Turns an id into a table name string"
+  [id]
+  (str table-prefix id))
+(def table-names (map id->table reg-range))
+
+(defn r
+  "Read a random subset of keys."
+  [_ _]
+  ;; Reading a random subset of keys after writing a random subset of
```
Comment (bdarnell) [inline cockroachdb/src/jepsen/cockroach/multiregister.clj]: Clarify this comment: is it describing what the code does now or a pitfall avoided? Does this mean that the test currently fails sometimes for this reason?

**Review by bdarnell — APPROVED** [review]

File: cockroachdb/src/jepsen/cockroach/multiregister.clj (hunk 1 above)
Comment (nvb) [reply cockroachdb/src/jepsen/cockroach/multiregister.clj]: I haven't seen the code fail yet for this reason, but I'm hoping to to confirm that things are working as expected. I'll address this comment before merging.

File: cockroachdb/src/jepsen/cockroach/runner.clj [hunk 2]
```diff
@@ -28,6 +29,7 @@
    "bank-multitable"      bank/multitable-test
    "comments"             comments/test
    "register"             register/test
+   "multiregister"        multiregister/test
```
Comment (ajwerner) [inline cockroachdb/src/jepsen/cockroach/runner.clj]: you called it "multi-register" in the README

File: cockroachdb/src/jepsen/cockroach/multiregister.clj [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    nil)
+
+  (close! [this test]
+    (rc/close! conn)))
+
+(defn test
+  [opts]
+  (cockroach/basic-test
+   (merge
+    {:name        "register"
```
Comment (ajwerner) [inline cockroachdb/src/jepsen/cockroach/multiregister.clj]: Should this be "multi-register"?

File: cockroachdb/src/jepsen/cockroach/multiregister.clj [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            [jepsen.cockroach.client :as c]
+            [jepsen.cockroach.nemesis :as cln]
+            [clojure.java.jdbc :as j]
+            [clojure.tools.logging :refer :all]
+            [knossos.model :as model]))
+
+(def reg-count 5)
+(def reg-range (range reg-count))
+
+(def table-prefix "String prepended to all table names." "register_")
```
Comment (ajwerner) [inline cockroachdb/src/jepsen/cockroach/multiregister.clj:24]: Do you want this to be different from the register test?

Comment (nvb) [reply cockroachdb/src/jepsen/cockroach/multiregister.clj:24]: It's fine if it differs. The register test should probably use a different table name but it doesn't really matter.

File: cockroachdb/src/jepsen/cockroach/multiregister.clj (hunk 3 above)
Comment (nvb) [reply cockroachdb/src/jepsen/cockroach/multiregister.clj]: Done.

File: cockroachdb/src/jepsen/cockroach/runner.clj (hunk 2 above)
Comment (nvb) [reply cockroachdb/src/jepsen/cockroach/runner.clj]: Done.

File: cockroachdb/src/jepsen/cockroach/multiregister.clj [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    (rc/close! conn)))
+
+(defn test
+  [opts]
+  (cockroach/basic-test
+   (merge
+    {:name         "multi-register"
+     :client      {:client (MultiAtomicClient. (atom false) nil)
+                   :during (independent/concurrent-generator
+                            (* 2 (count (:nodes opts)))
```
Comment (ajwerner) [inline cockroachdb/src/jepsen/cockroach/multiregister.clj]: Do you want to comment these values? They appear to configure the workload behavior but I don't know what they actually do. Maybe that's okay

**Review by ajwerner — APPROVED** [review]
LGTM with a request for some comments on the test definition

File: cockroachdb/src/jepsen/cockroach/multiregister.clj (hunk 5 above)
Comment (nvb) [reply cockroachdb/src/jepsen/cockroach/multiregister.clj]: They're like this for every test (for instance, [here](https://github.com/cockroachdb/jepsen/blob/master/cockroachdb/src/jepsen/cockroach/register.clj#L93)), so I'm inclined to keep this as is.

**Pull request merged**